src/Microsoft.ML.StandardTrainers/LdSvm/LdSvmTrainer.cs

```diff
@@ -0,0 +1,489 @@
+            ///  Parameter for sigmoid sharpness
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Parameter for sigmoid sharpness", ShortName = "s", SortOrder = 50)]
+            [TGUI(SuggestedSweeps = "1.0,0.1,0.01")]
+            [TlcModule.SweepableDiscreteParam("Sigma", new object[] { 1.0f, 0.1f, 0.01f })]
+            public float Sigma = Defaults.Sigma;
+
+            /// <summary>
+            /// Indicates if we should use Bias or not in our model.
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "No bias", ShortName = "noBias")]
+            [TlcModule.SweepableDiscreteParam("NoBias", null, isBool: true)]
+            public bool NoBias = Defaults.NoBias;
+
+            /// <summary>
+            /// Number of iterations
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce,
+                HelpText = "Number of iterations", ShortName = "iter,NumIterations", SortOrder = 50)]
+            [TGUI(SuggestedSweeps = "10000,15000")]
+            [TlcModule.SweepableDiscreteParam("NumIterations", new object[] { 10000, 15000 })]
+            public int NumberOfIterations = Defaults.NumberOfIterations;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "The calibrator kind to apply to the predictor. Specify null for no calibration", Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly)]
+            internal ICalibratorTrainerFactory Calibrator = new PlattCalibratorTrainerFactory();
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "The maximum number of examples to use when training the calibrator", Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly)]
+            public int MaxCalibrationExamples = 1000000;
+
+            internal class Defaults
+            {
+                public const int NumberOfIterations = 15000;
+                public const bool NoBias = false;
+                public const float Sigma = 1.0f;
+                public const float LambdaThetaprime = 0.01f;
+                public const float LambdaTheta = 0.01f;
+                public const float LambdaW = 0.1f;
+                public const int TreeDepth = 3;
+            }
+        }
+
+        private Options _options;
+
+        internal LdSvmTrainer(IHostEnvironment env, Options options)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(LoadNameValue),
+                  TrainerUtils.MakeR4VecFeature(options.FeatureColumnName),
+                  TrainerUtils.MakeBoolScalarLabel(options.LabelColumnName),
+                  TrainerUtils.MakeR4ScalarWeightColumn(options.ExampleWeightColumnName))
+        {
+            Host.CheckValue(options, nameof(options));
+            CheckOptions(Host, options);
+            _options = options;
+        }
+
+        // REVIEW: This does not need caching, but only because it's very
+        // badly written and the first thing is it just grabs all instances. If this
```

Comment:
> `badly written and the first thing is it just grabs all instances`

Does this mean this trainer can't stream data, but instead it needs to load all the data into memory before it can train on it? Is that acceptable?

Reply:
When looking thru the code, it seemed like a sizable amount of work to bring this to a streaming mode. I didn't see an obvious method to be performant and streaming -- seems like a lots of passes over the dataset. I very well may be missing the obvious though.

Reply:
In the most recent iteration I added two code paths: one does streaming and one does caching. From running some experiments, the streaming version is around 2x slower. Could you take a look and let me know whether it's worth keeping both?

@justinormont, @harishsk

---
In reply to: [313606773](https://github.com/dotnet/machinelearning/pull/4060#discussion_r313606773) [](ancestors = 313606773)

Reply:
Is the streaming version 2x slower even in the case of large data? If there is a real tradeoff for large data sizes, then it is worth keeping both. 

---
In reply to: [368572412](https://github.com/dotnet/machinelearning/pull/4060#discussion_r368572412) [](ancestors = 368572412,313606773)